spec/formatting.md

```diff
@@ -1,24 +1,124 @@
 
-The document is part of the MessageFormat 2.0 specification,
-the successor to ICU MessageFormat, henceforth called ICU MessageFormat 1.0.
+To start, we presume that a _message_ has either been parsed from its syntax
+or created from a data model description.
+If this construction has encountered any Syntax or Data Model errors,
+their handling during formatting is specified here as well.
 
-## Literal Resolution
+Formatting of a _message_ is defined by the following operations:
 
-The resolved value of a _text_ or a _literal_ is the character sequence of the _text_
-or _literal_ after any character escape has been converted to the escaped character.
-When a _literal_ is used as an _expression_ argument or on the right-hand side of an _option_,
-the formatting function MUST treat their resolved values the same independently of whether the
-value was originally _quoted_ or _unquoted_.
-For example, the _option_ `foo=42` and the _option_ `foo=|42|` 
-are treated as identical.
+- **_Expression Resolution_** determines the value of an _expression_,
+  with reference to the current _formatting context_.
+  This can include multiple steps,
+  such as looking up the value of a variable and calling formatting functions.
+  The resolved value is not necessarily in the shape it will finally take,
+  but is "formattable", i.e. it contains everything required by the eventual formatting.
 
-## Variable Resolution
+  The resolution of _text_ is rather straighforward,
+  and is detailed under _literal resolution_.
+
+- **_Pattern Selection_** determines which of a message's _patterns_ is formatted.
+  For a message with no _selectors_, this is simple as there is only one _pattern_.
+  With _selectors_, this will depend on their resolution.
+
+- **_Formatting_** takes the resolved values of the selected _pattern_,
+  and produces the formatted result for the _message_.
+  Depending on the implementation, this result could be a single concatenated string,
+  an array of objects, an attributed string, or some other locally appropriate data type.
+
+Formatter implementations are not required to expose
+the _expression resolution_ and _pattern selection_ operations to their users,
+or even use them in their internal processing,
+as long as the final _formatting_ result is made available to users
+and the observable behavior of the formatter matches that described here.
+
+## Formatting Context
+
+A message's **_formatting context_** represents the data and procedures that are required
+for the message's _expression resolution_, _pattern selection_ and _formatting_.
+
+At a minimum, it includes:
+
+- Information on the current **_locale_**,
+  potentially including a fallback chain of locales.
+  This will be passed on to formatting functions.
+
+- Information on the base directionality of the message and its _text_ tokens.
+  This will be used by strategies for bidirectional isolation.
```

Comment:
Not just! It should also be exposed as a property of the message so that the caller can properly set the base paragraph direction. Or is this what you mean by "strategies for bidirectional isolation"? If the latter, I'd suggest:

```suggestion
This can be used by callers to set the base direction of the _message_ upon display.
```

Reply:
Included in #411.